cirq-core/cirq/transformers/routing/mapping_manager.py

```diff
@@ -0,0 +1,113 @@
+
+from cirq import ops, protocols
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class MappingManager:
+    """Class that keeps track of the mapping of logical to physical qubits and provides
+    convenience methods for distance queries on the physical qubits.
+
+    Qubit variables with the characer 'p' preppended to them are physical and qubits with  the
+    character 'l' preppended to them are logical qubits.
+    """
+
+    def __init__(self, device_graph: nx.Graph, initial_mapping: Dict[ops.Qid, ops.Qid]) -> None:
+        """Initializes MappingManager.
+
+        Args:
+            device_graph: connectivity graph of qubits in the hardware device.
+            circuit_graph: connectivity graph of the qubits in the input circuit.
+            initial_mapping: the initial mapping of logical (keys) to physical qubits (values).
+        """
+        self.device_graph = device_graph
+        self._map = initial_mapping.copy()
+        self._inverse_map = {v: k for k, v in self._map.items()}
+        self._induced_subgraph = nx.induced_subgraph(self.device_graph, self._map.values())
+        self._shortest_paths_matrix = dict(nx.all_pairs_shortest_path(self._induced_subgraph))
+
+    @property
+    def map(self) -> Dict[ops.Qid, ops.Qid]:
+        """The mapping of logical qubits (keys) to physical qubits (values)."""
+        return self._map
+
+    @property
+    def inverse_map(self) -> Dict[ops.Qid, ops.Qid]:
+        """The mapping of physical qubits (keys) to logical qubits (values)."""
+        return self._inverse_map
+
+    @property
+    def induced_subgraph(self) -> nx.Graph:
+        """The device_graph induced on the physical qubits that are mapped to."""
+        return self._induced_subgraph
+
+    def dist_on_device(self, lq1: ops.Qid, lq2: ops.Qid) -> int:
+        """Finds shortest path distance path between the corresponding physical qubits for logical
+        qubits q1 and q2 on the device.
+
+        Args:
+            lq1: the first logical qubit.
+            lq2: the second logical qubit.
+
+        Returns:
+            The shortest path distance.
+        """
+        return len(self._shortest_paths_matrix[self._map[lq1]][self._map[lq2]]) - 1
+
+    def can_execute(self, op: ops.Operation) -> bool:
+        """Finds whether the given operation can be executed on the device.
+
+        Args:
+            op: an operation on logical qubits.
+
+        Returns:
+            Whether the given operation is executable on the device.
+        """
+        return protocols.num_qubits(op) < 2 or self.dist_on_device(*op.qubits) == 1
+
+    def apply_swap(self, lq1: ops.Qid, lq2: ops.Qid) -> None:
+        """Swaps two logical qubits in the map and in the inverse map.
+
+        Args:
+            lq1: the first logical qubit.
+            lq2: the second logical qubit.
+        """
+        self._map[lq1], self._map[lq2] = self._map[lq2], self._map[lq1]
+
+        pq1 = self._map[lq1]
+        pq2 = self._map[lq2]
```

Comment:
It doesn't matter because you are just swapping, but ideally `pq1` and `pq2` should be computed before `self._map[lq1]` and `self._map[lq2]` have been swapped. 
```suggestion
        pq1, pq2 = self._map[lq1], self._map[lq2]
```